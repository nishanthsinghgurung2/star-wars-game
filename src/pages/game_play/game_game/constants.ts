import { STARSHIPS_CARD_IMGS, PEOPLE_CARD_IMGS } from './graphics';

export interface CardSuitMeta {
  label: string;
  path: string;
  img: string;
}

export const BASE_SUIT_PATH = '/boards';

export const CARD_SUIT_META = {
  starships: {
    label: 'Starships',
    path: BASE_SUIT_PATH + '/starships',
    img:
      'https://scandinaviantraveler.com/sites/default/files/styles/article_top_full_imagecustom_user_desktop_large_1x/public/starwars-fleet-1140x619.jpg?itok=jDZnk9oX',
    cardImgs: STARSHIPS_CARD_IMGS,
  },
  people: {
    label: 'People',
    path: BASE_SUIT_PATH + '/people',
    img: 'https://pvpantherprint.org/wp-content/uploads/2018/03/star-wars-prequels.jpg',
    cardImgs: PEOPLE_CARD_IMGS,
  },
};
