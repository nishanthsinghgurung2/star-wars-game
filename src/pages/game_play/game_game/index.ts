import { Entity } from '../../../client/client_types';

export const deal = (cards: Entity[], players: number): Entity[] => {
  if (players < 2 || players > cards.length)
    throw RangeError('Invalid number of players for deck size. Must be 2 < players < cards.length');
  return Array(players)
    .fill(0)
    .flatMap(() => {
      const ranIndex = Math.floor(Math.random() * (cards.length - 1));
      return cards.splice(ranIndex, 1);
    });
};

export const checkWinner = <T>(
  hands: T[],
  compField: keyof T,
): { index: number; value: number } => {
  if (hands.length < 2) throw RangeError('hands argument is not valid. Must be > 2');
  const values = hands.map((card) => parseFloat(card[compField]?.toString()) || 0);
  const value = Math.max(...values);
  const index = values.findIndex((cval) => cval === value);
  return { index, value };
};
