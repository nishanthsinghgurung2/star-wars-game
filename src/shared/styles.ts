import styled from 'styled-components';
import { Container, VariantContainer } from '../styles/styles';

export const LayoutGrid = styled(VariantContainer)`
  background: ${(props) => props.theme.gridBackground};
  color: ${(props) => props.theme.color};
  font-family: 'Noto Sans', sans-serif;

  display: grid;
  height: 100%;
  grid-template-areas:
    'header'
    'main';
  grid-template-rows: 100px minmax(100px, auto);
`;

export const RowContainer = styled(Container)`
  flex-direction: row;
  justify-content: space-around;
`;

export const HighContainer = styled(VariantContainer)`
  height: 100%;
`;

export const Header = styled(Container)`
  grid-area: header;
`;

export const Main = styled(VariantContainer)`
  grid-area: main;
  height: 100%;
`;
