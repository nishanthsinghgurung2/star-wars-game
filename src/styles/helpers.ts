export const camelCaseToSentenceCase = (camelCaseString: string): string => {
  if (camelCaseString === '' || camelCaseString == null) return '';
  const result = camelCaseString.replace(/([A-Z])/g, ' $1');
  return result.charAt(0).toUpperCase() + result.slice(1);
};
