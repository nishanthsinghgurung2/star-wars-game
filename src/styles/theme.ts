export interface IPalette {
  main: string;
  contrastText: string;
}

export interface Theme {
  borderRadius: string;
  color: string;
  background: string;
  palette: {
    dark: IPalette;
    light: IPalette;
    primary: IPalette;
    secondary: IPalette;
    success: IPalette;
    info: IPalette;
    warning: IPalette;
    danger: IPalette;
  };
}
// https://loading.io/color/feature/
export const defaultTheme: Theme = {
  borderRadius: '4px',
  color: '#FFF',
  background: '#0E0C0F',
  palette: {
    dark: {
      main: '#0E0C0F',
      contrastText: '#ffffff',
    },
    light: {
      main: '#6a737b',
      contrastText: '#ffffff',
    },
    primary: {
      main: '#00a4e4',
      contrastText: '#ffffff',
    },
    secondary: {
      main: '#8a7967',
      contrastText: '#ffffff',
    },
    success: {
      main: '#c1d82f',
      contrastText: '#ffffff',
    },
    info: {
      main: '#ffdd00',
      contrastText: '#ffffff',
    },
    warning: {
      main: '#fbb034',
      contrastText: '#ffffff',
    },
    danger: {
      main: '#ff0000',
      contrastText: '#ffffff',
    },
  },
};

export const secondaryTheme: Theme = {
  borderRadius: '4px',
  color: '#FFF',
  background: '#0E0C0F',
  palette: {
    dark: {
      main: '#52565e',
      contrastText: '#ffffff',
    },
    light: {
      main: '#caccd1',
      contrastText: '#ffffff',
    },
    primary: {
      main: '#0a8ea0',
      contrastText: '#ffffff',
    },
    secondary: {
      main: '#30c39e',
      contrastText: '#ffffff',
    },
    success: {
      main: '#30c39e',
      contrastText: '#ffffff',
    },
    info: {
      main: '#ffc845',
      contrastText: '#ffffff',
    },
    warning: {
      main: '#f48924',
      contrastText: '#ffffff',
    },
    danger: {
      main: '#f85a40',
      contrastText: '#ffffff',
    },
  },
};
