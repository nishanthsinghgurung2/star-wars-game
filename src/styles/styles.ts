import styled from 'styled-components';
import { IPalette, Theme } from './theme';

export interface IVariant {
  variant?: string;
  theme: Theme;
}

export const vaiantPalette = (props: IVariant): IPalette =>
  props.theme?.palette?.[props.variant] || props.theme?.palette?.dark;

export const VariantContainer = styled.div<IVariant>`
  background: ${(props) => vaiantPalette(props)?.main};
  color: ${(props) => vaiantPalette(props)?.contrastText};
`;

export const Container = styled(VariantContainer)`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  font-size: 2rem;
  box-shadow: rgba(3, 8, 20, 0.1) 0px 0.15rem 0.5rem, rgba(2, 8, 20, 0.1) 0px 0.075rem 0.175rem;
  height: 100%;
  width: 100%;
  transition: all 500ms;
  overflow: hidden;

  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;

  &.active,
  &:hover {
    color: ${(props) => props.theme?.palette?.primary?.main};
  }
`;

export const Navbar = styled.div``;
