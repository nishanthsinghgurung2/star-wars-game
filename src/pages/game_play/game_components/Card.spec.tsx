import '@testing-library/jest-dom/extend-expect';
import { render, screen, within } from '@testing-library/react';
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { CARD_SUIT_META } from '../game_game/constants';
import { entityFactories, entityStaticFactories } from '../../../test_utils/facroties';
import { camelCaseToSentenceCase } from '../../../styles/helpers';
import Card, { IDENTITY_FIELDS } from './game_card';

describe('<Card />', () => {
  describe.each(Object.entries(CARD_SUIT_META))('for type (%s)', (key, val) => {
    it('renders correctly', () => {
      const card = entityStaticFactories[key].build();

      const { container } = render(
        <MemoryRouter initialEntries={[val.path]}>
          <Card card={card} />
        </MemoryRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should show humanly formated field name and field value correctly', () => {
      const card = entityFactories[key].build();

      render(
        <MemoryRouter initialEntries={[val.path]}>
          <Card card={card} />
        </MemoryRouter>,
      );

      const cardUi = screen.getByTestId(`card-test-id-${card.id}`);

      expect(within(cardUi).queryByText(card.id)).not.toBeInTheDocument();
      expect(within(cardUi).getByText(card.name)).toBeInTheDocument();

      Object.entries(card)
        .filter(([key]) => !IDENTITY_FIELDS.includes(key))
        .map(([key, val]) => {
          expect(within(cardUi).getByText(camelCaseToSentenceCase(key))).toBeInTheDocument();
          expect(within(cardUi).getByText(val as string)).toBeInTheDocument();
        });
    });
  });
});
