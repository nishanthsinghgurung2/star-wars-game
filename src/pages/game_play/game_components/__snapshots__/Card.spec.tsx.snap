// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> for type (people) renders correctly 1`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz sc-jSgupP emDVYV iPcSwN IfzQm"
  data-testid="card-test-id-666"
>
  <div
    class="sc-bdfBwQ sc-gsTCUz sc-gKsewC emDVYV iPcSwN kXWqov"
  >
    jedi
  </div>
  <div
    class="sc-bdfBwQ sc-gsTCUz sc-iBPRYJ emDVYV iPcSwN fJnjGu"
    data-testid="card-body-id-666"
  >
    <div
      class="sc-bdfBwQ sc-gsTCUz sc-fubCfw emDVYV iPcSwN jPncyk"
    >
      <label
        class="sc-pFZIQ"
      >
        Birth Year
      </label>
      <span
        class="sc-jrAGrp iBfGkU"
      >
        11BB
      </span>
    </div>
    <div
      class="sc-bdfBwQ sc-gsTCUz sc-fubCfw emDVYV iPcSwN jPncyk"
    >
      <label
        class="sc-pFZIQ"
      >
        Mass
      </label>
      <span
        class="sc-jrAGrp iBfGkU"
      >
        81.5
      </span>
    </div>
    <div
      class="sc-bdfBwQ sc-gsTCUz sc-fubCfw emDVYV iPcSwN jPncyk"
    >
      <label
        class="sc-pFZIQ"
      >
        Height
      </label>
      <span
        class="sc-jrAGrp iBfGkU"
      >
        175
      </span>
    </div>
  </div>
</div>
`;

exports[`<Card /> for type (starships) renders correctly 1`] = `
<div
  class="sc-bdfBwQ sc-gsTCUz sc-jSgupP emDVYV iPcSwN IfzQm"
  data-testid="card-test-id-12345"
>
  <div
    class="sc-bdfBwQ sc-gsTCUz sc-gKsewC emDVYV iPcSwN kXWqov"
  >
    jedi star fighter
  </div>
  <div
    class="sc-bdfBwQ sc-gsTCUz sc-iBPRYJ emDVYV iPcSwN fJnjGu"
    data-testid="card-body-id-12345"
  >
    <div
      class="sc-bdfBwQ sc-gsTCUz sc-fubCfw emDVYV iPcSwN jPncyk"
    >
      <label
        class="sc-pFZIQ"
      >
        Model
      </label>
      <span
        class="sc-jrAGrp iBfGkU"
      >
        delta-7 model
      </span>
    </div>
    <div
      class="sc-bdfBwQ sc-gsTCUz sc-fubCfw emDVYV iPcSwN jPncyk"
    >
      <label
        class="sc-pFZIQ"
      >
        Length
      </label>
      <span
        class="sc-jrAGrp iBfGkU"
      >
        10
      </span>
    </div>
    <div
      class="sc-bdfBwQ sc-gsTCUz sc-fubCfw emDVYV iPcSwN jPncyk"
    >
      <label
        class="sc-pFZIQ"
      >
        Cargo Capacity
      </label>
      <span
        class="sc-jrAGrp iBfGkU"
      >
        1000
      </span>
    </div>
    <div
      class="sc-bdfBwQ sc-gsTCUz sc-fubCfw emDVYV iPcSwN jPncyk"
    >
      <label
        class="sc-pFZIQ"
      >
        Hyperdrive Rating
      </label>
      <span
        class="sc-jrAGrp iBfGkU"
      >
        1.5
      </span>
    </div>
  </div>
</div>
`;
